gitlab/resource_gitlab_project_environment.go

```diff
@@ -0,0 +1,163 @@
+				ForceNew:     true,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			"external_url": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			"state": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceGitlabProjectEnvironmentCreate(d *schema.ResourceData, meta interface{}) error {
+	name := d.Get("name").(string)
+	options := gitlab.CreateEnvironmentOptions{
+		Name: &name,
+	}
+	if externalURL, ok := d.GetOk("external_url"); ok {
+		options.ExternalURL = gitlab.String(externalURL.(string))
+	}
+
+	project := d.Get("project").(string)
+
+	log.Printf("[DEBUG] Project %s create gitlab environment %q", project, *options.Name)
+
+	client := meta.(*gitlab.Client)
+
+	environment, resp, err := client.Environments.CreateEnvironment(project, &options)
+	if err != nil {
+		if resp.StatusCode == http.StatusNotFound {
+			return fmt.Errorf("feature Environments is not available")
+		}
+		return err
+	}
+
+	d.SetId(buildTwoPartID(&project, gitlab.String(fmt.Sprintf("%v", environment.ID))))
+
+	return resourceGitlabProjectEnvironmentRead(d, meta)
+}
+
+func resourceGitlabProjectEnvironmentRead(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[DEBUG] read gitlab environment %s", d.Id())
+
+	project, environmentID, err := parseTwoPartID(d.Id())
+	if err != nil {
+		return err
+	}
+	environmentIDInt, err := strconv.Atoi(environmentID)
+
+	log.Printf("[DEBUG] Project %s read gitlab environment %d", project, environmentIDInt)
+
+	client := meta.(*gitlab.Client)
+
+	environment, resp, err := client.Environments.GetEnvironment(project, environmentIDInt)
+	if err != nil {
+		if resp.StatusCode == http.StatusNotFound {
+			log.Printf("[DEBUG] Project %s gitlab environment %q not found", project, environmentID)
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("error getting gitlab project %q environment %q: %v", project, environmentID, err)
+	}
+
+	d.SetId(buildTwoPartID(&project, gitlab.String(fmt.Sprintf("%v", environment.ID))))
+	d.Set("project", project)
+	d.Set("name", environment.Name)
+	d.Set("state", environment.State)
+
+	return nil
+}
+
+func resourceGitlabProjectEnvironmentUpdate(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[DEBUG] update gitlab environment %s", d.Id())
+
+	project, environmentID, err := parseTwoPartID(d.Id())
+	if err != nil {
+		return err
+	}
+	environmentIDInt, err := strconv.Atoi(environmentID)
+	if err != nil {
+		return fmt.Errorf("error converting environment ID to int: %v", err)
+	}
+
+	name := d.Get("name").(string)
+	options := gitlab.EditEnvironmentOptions{
+		Name: &name,
+	}
+	if externalURL, ok := d.GetOk("external_url"); ok {
+		options.ExternalURL = gitlab.String(externalURL.(string))
```

Comment:
Usually use HasChange in the update function

```suggestion
	if d.HasChange("external_url") {
		options.ExternalURL = gitlab.String(d.Get("external_url").(string))
```